### PR TITLE
test(pgwire): infrastructure to run Python and Rust tests

### PR DESCRIPTION
**Pull request by jerrinot**
This changeset adds a GitHub action with Python psycopg2, psycopg3, asyncpg runners for PGWire YAML scenarios. 
It also adds a Rust runner which uses tokio-postgres clients. Running the same scenarios as Python runners. 
  
It's a foundation for testing all popular (non-Java) clients.

### CODE REVIEW

**Comment by questdb-butler** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true">Fails</th>
    </tr>
  </thead>
  <tbody><tr>
      <td>:no_entry_sign:</td>
      <td>

  Please update the PR title to match this format:
`type(subType): description`

Where `type` is one of:
`feat`, `fix`, `chore`, `docs`, `style`, `refactor`, `perf`, `test`, `ci`, `revert`

And: `subType` is one of:
`build`, `sql`, `log`, `mig`, `core`, `ilp`, `pgwire`, `http`, `conf`, `ui`, `wal`

For Example:

```
perf(sql): improve pattern matching performance for SELECT sub-queries
```
  </td>
    </tr>
  </tbody>
</table>




<p align="right">
  Generated by :no_entry_sign: <a href="https://danger.systems/js">dangerJS</a> against 2fc6e26d69980f97f492d4453722715afc9849fc
</p>

**Comment by ideoma** [comment]
Can we use sqllogictest framework to run these tests?
It's already uses Rust client to execute the queries.

**Comment by jerrinot** [comment]
@ideoma: It's complicated. This PR is introducing two kinds of tests:
- Shared Tests - Described in YAML and executed by all runners. Some of these tests already exist, with more to come.
- Client-Specific Tests - Will push them today. 

### Shared Tests
SQLLogic is similar to  no. 1: There is an abstract representation of a test scenario, which is interpreted by a client-specific runner. This is quite similar to what I’ve been working on. I wish I had realized this on Monday. :(

There’s a SQLLogic runner in Rust and Java (JDBC), but that’s it? No runner for Python, C#, Node.js, or R? This means we’ll need to implement SQLLogic runners for each language. You could argue we also need to implement our own  language-specific YAML test runner, but I suspect the SQLLogic runner is more complex than our YAML runners. Some data points: our Python runners have around 200 lines of code (LoC), our Rust YAML runner has 450 LoC, while the SQLLogic Rust runner has 3.3k LoC. So, it’s probably more complicated. Still, there may be valid reasons for the extra complexity, and the jury is still out on whether creating our own test-definition language was the right decision (no test re-use 😞).


### Client-Specific Tests
While implementing YAML runners for #1, I realized that an abstract format is quite limiting. I can only test what is supported by all clients. I can do white/blacklisting, but it’s still not enough to fully exercise the clients. For example, some clients explicitly support pgwire pipelining and may submit multiple queries; some support different pagination styles; some provide an API to describe type parameters in user-supplied SQL (e.g., SELECT * FROM foo WHERE bar = ?), there are different programming models (sync vs async), etc.

Here’s how I see it:
-  YAML runners are client protocol compatibility smoke tests. You add a query to a YAML file, and all runners will pick it up. It’s relatively simple to code a new runner, but we have to write all the tests ourselves.
-  Client-specific tests are advanced client compatibility tests. They allow us to exercise more complex, client-specific features.
-  SQLLogic is a SQL correctness test. Implementing a new runner is non-trivial, but there are existing third-party tests that can be reused.

Besides this, we should also have another kind of tests: low-level pgwire compliance tests. Based not on any client calls, but on explicit message exchange. This is required for testing edge-cases, QuestDB behaviour during protocol violation, etc. Similar to what eg. CockroachDB has [here](https://github.com/cockroachdb/cockroach/blob/v22.1.3/pkg/sql/pgwire/testdata/pgtest/bind_and_resolve).

File: compat/src/test/python/runner_psycopg3.py [hunk 1]
```diff
@@ -0,0 +1,161 @@
...
+import sys
+import yaml
+from psycopg import Connection, Cursor
+from string import Template
+from common import *
+
+
+def adjust_placeholder_syntax(query):
+    # Replace $[n] with %s
+    return re.sub(r'\$\[\d+\]', '%s', query)
```
[the remark this answers is not in the document]
Comment (mtopolnik) [reply compat/src/test/python/runner_psycopg3.py:32]: This function is specific to each library.

**Pull request merged**